org.eclipse.jdt.core.javac/src/org/eclipse/jdt/core/dom/JavacCompilationUnitResolver.java

```diff
@@ -125,29 +229,36 @@ public CompilationUnit toCompilationUnit(org.eclipse.jdt.internal.compiler.env.I
 
 	public CompilationUnit parse(org.eclipse.jdt.internal.compiler.env.ICompilationUnit sourceUnit, int apiLevel, Map<String, String> compilerOptions,
 			int flags, IJavaProject javaProject, IProgressMonitor monitor) {
+		Context context = new Context();
+		JavacUtils.configureJavacContext(context, compilerOptions, javaProject);
+		return parse(sourceUnit, apiLevel, compilerOptions, flags, javaProject, monitor, context);
+	}
+
+	private CompilationUnit parse(org.eclipse.jdt.internal.compiler.env.ICompilationUnit sourceUnit, int apiLevel, Map<String, String> compilerOptions,
+			int flags, IJavaProject javaProject, IProgressMonitor monitor, Context context) {
 		SimpleJavaFileObject fileObject = new SimpleJavaFileObject(new File(new String(sourceUnit.getFileName())).toURI(), JavaFileObject.Kind.SOURCE) {
 			@Override
 			public CharSequence getCharContent(boolean ignoreEncodingErrors) throws java.io.IOException {
 				return new String(sourceUnit.getContents());
 			}
 		};
-		Context context = new Context();
 		AST ast = createAST(compilerOptions, apiLevel, context);
 //		int savedDefaultNodeFlag = ast.getDefaultNodeFlag();
 //		ast.setDefaultNodeFlag(ASTNode.ORIGINAL);
 //		ast.setDefaultNodeFlag(savedDefaultNodeFlag);
 		ast.setDefaultNodeFlag(ASTNode.ORIGINAL);
 		CompilationUnit res = ast.newCompilationUnit();
-		context.put(DiagnosticListener.class, diagnostic -> {
-			if (Objects.equals(diagnostic.getSource(), fileObject) ||
-				diagnostic.getSource() instanceof DiagnosticSource source && Objects.equals(source.getFile(), fileObject)) {
-				IProblem[] previous = res.getProblems();
-				IProblem[] newProblems = Arrays.copyOf(previous, previous.length + 1);
-				newProblems[newProblems.length - 1] = JavacConverter.convertDiagnostic(diagnostic);
-				res.setProblems(newProblems);
-			}
-		});
-		JavacUtils.configureJavacContext(context, compilerOptions, javaProject);
+		if (context.get(DiagnosticListener.class) == null) {
```

Comment:
1 issue can be (I'm not sure how JDT currently behaves) that if you ask `resolve` for a set of files, the errors would be reported on all files, or at least on each AST.
In the current form, the DiagnosticListener will only report errors about the 1st file. We most likely need a smarter DiagnosticResolver that is able to work for all files in the context. Probably we should add 1 diagnostic listener that would store a `Map<JavaFileObject, CompilationUnit>` and we could register in this diagnosticlistener the various parsed files so diagnostics get routed to the right unit.

Note that I recently refactored this part a bit. If you want, I can take care of the rebasing and adaptation here.